### DISCUSSION TITLE
Discrepancy Between Objective Flux and Biomass Flux in pFBA and loopless_solution

### DISCUSSION THREAD

**Discussion by AmalReji** [opencobra/cobrapy#1349] [Q&A] [answered]
Hi everyone, I am encountering an issue with CobraPy's pFBA function. The reported objective flux from `pFBA(model)` does not match the calculated biomass flux. I also noticed the same issue with `loopless_solution(model)`, but `model.optimize()` seems to work fine.

**Example script**
```python
model = cobra.io.sbml.read_sbml_model(path_to_sbml_model)
model.objective = original_model.reactions.get_by_id('Biomass')
solution = model.optimize()
p_solution = pfba(model)
l_solution = loopless_solution(model)

print("Objective Flux (normal):", solution.objective_value)
print("Biomass Flux (normal):", solution.fluxes["Biomass"])
print("Objective Flux (pfba):", p_solution.objective_value)
print("Biomass Flux (pfba):", p_solution.fluxes["Biomass"])
print("Objective Flux (loopless):", p_solution.objective_value)
print("Biomass Flux (loopless):", p_solution.fluxes["Biomass"])
```

**Output**
```
Objective Flux (normal): 15.97723933580795
Biomass Flux (normal): 15.97723933580795

Objective Flux (pfba): 7708.576072346518
Biomass Flux (pfba): 15.977239335807882

Objective Flux (loopless): 0.0
Biomass Flux (loopless): 15.97723933580795
```

My Question: Why is objective flux != biomass flux when using `pFBA(model)` and `loopless_solution(model)`, but equal when using `model.optimize()`?

If any additional information or code is needed to better understand the problem, please let me know. Thank you for your time and assistance!

**Comment by cdiener** [comment, marked as answer]
In pfba the objective is the sum of absolute fluxes, so that is what is returned. 

For the loopless solution that is a known bug and should be fixed when #1341 is merged.

**Reply by AmalReji** [reply]
Thank you! Did you know if CobraPy had any other optimisation algorithms which enforce thermodynamic feasibility? These are all the algorithms I've come across so far: FBA, pFBA, loopless_solution and geometric_FBA.

**Reply by Midnighter** [reply]
You would probably have to combine equilibrator (https://equilibrator.readthedocs.io/) to set thermodynamic constraints.
